modules/mig/main.tf

```diff
@@ -15,85 +15,162 @@
  */
 
 locals {
-  distribution_policy_zones = {
-    default = ["${data.google_compute_zones.available.names}"]
-    user    = ["${var.distribution_policy_zones}"]
+  distribution_policy_zones_base = {
+    default = data.google_compute_zones.available.names
+    user    = var.distribution_policy_zones
   }
+  distribution_policy_zones = local.distribution_policy_zones_base[length(var.distribution_policy_zones) == 0 ? "default" : "user"]
 }
 
-data "google_compute_zones" "available" {}
+data "google_compute_zones" "available" {
+}
 
 resource "google_compute_region_instance_group_manager" "mig" {
-  provider           = "google-beta"
-  base_instance_name = "${var.hostname}"
+  provider           = google-beta
+  base_instance_name = var.hostname
 
-  version = {
+  version {
     name              = "${var.hostname}-mig-version-0"
-    instance_template = "${var.instance_template}"
+    instance_template = var.instance_template
   }
 
-  name         = "${var.hostname}-mig"
-  region       = "${var.region}"
-  named_port   = ["${var.named_ports}"]
-  target_pools = ["${var.target_pools}"]
-  target_size  = "${var.autoscaling_enabled ? var.min_replicas : var.target_size}"
+  name   = "${var.hostname}-mig"
+  region = var.region
+  dynamic "named_port" {
+    for_each = var.named_ports
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
+      # which keys might be set in maps assigned here, so it has
+      # produced a comprehensive set here. Consider simplifying
+      # this after confirming which keys can be set in practice.
+
+      name = lookup(named_port.value, "name", null)
+      port = lookup(named_port.value, "port", null)
+    }
+  }
+  target_pools = var.target_pools
+  target_size  = var.autoscaling_enabled ? var.min_replicas : var.target_size
 
   auto_healing_policies {
-    health_check      = "${var.http_healthcheck_enable || var.tcp_healthcheck_enable ? element(concat(google_compute_health_check.http_healthcheck.*.self_link, google_compute_health_check.tcp_healthcheck.*.self_link, list("")),0) : ""}"
-    initial_delay_sec = "${var.hc_initial_delay_sec}"
+    health_check = var.http_healthcheck_enable || var.tcp_healthcheck_enable ? element(
+      concat(
+        google_compute_health_check.http_healthcheck.*.self_link,
+        google_compute_health_check.tcp_healthcheck.*.self_link,
+        [""],
+      ),
+      0,
+    ) : ""
```

Comment:
Please check if "null" is better

Reply:
I'll check

Reply:
@ingwarr did you determine if `null` is more appropriate?